crml/staking/src/lib.rs

```diff
@@ -1565,98 +2264,447 @@ impl<T: Trait> Module<T> {
+	///
+	/// State reads: ElectionState, CurrentEr, QueuedScore.
+	///
+	/// This function does weight refund in case of errors, which is based upon the fact that it is
+	/// called at the very beginning of the call site's function.
+	pub fn pre_dispatch_checks(score: ElectionScore, era: EraIndex) -> DispatchResultWithPostInfo {
+		// discard solutions that are not in-time
+		// check window open
+		ensure!(
+			Self::era_election_status().is_open(),
+			Error::<T>::OffchainElectionEarlySubmission.with_weight(T::DbWeight::get().reads(1)),
 		);
 
-		if let Some(phragmen_result) = maybe_phragmen_result {
-			let elected_stashes = phragmen_result
-				.winners
-				.into_iter()
-				.map(|(s, _)| s)
-				.collect::<Vec<T::AccountId>>();
-			let assignments = phragmen_result.assignments;
+		// check current era.
+		if let Some(current_era) = Self::current_era() {
+			ensure!(
+				current_era == era,
+				Error::<T>::OffchainElectionEarlySubmission.with_weight(T::DbWeight::get().reads(2)),
+			)
+		}
 
-			let staked_assignments =
-				sp_npos_elections::assignment_ratio_to_staked(assignments, Self::slashable_balance_of_vote_weight);
+		// assume the given score is valid. Is it better than what we have on-chain, if we have any?
+		if let Some(queued_score) = Self::queued_score() {
+			ensure!(
+				is_score_better(score, queued_score, T::MinSolutionScoreBump::get()),
+				Error::<T>::OffchainElectionWeakSubmission.with_weight(T::DbWeight::get().reads(3)),
+			)
+		}
+
+		Ok(None.into())
+	}
+
+	/// Checks a given solution and if correct and improved, writes it on chain as the queued result
+	/// of the next round. This may be called by both a signed and an unsigned transaction.
+	pub fn check_and_replace_solution(
+		winners: Vec<ValidatorIndex>,
+		compact_assignments: CompactAssignments,
+		compute: ElectionCompute,
+		claimed_score: ElectionScore,
+		era: EraIndex,
+		election_size: ElectionSize,
+	) -> DispatchResultWithPostInfo {
+		// Do the basic checks. era, claimed score and window open.
+		let _ = Self::pre_dispatch_checks(claimed_score, era)?;
+
+		// before we read any further state, we check that the unique targets in compact is same as
+		// compact. is a all in-memory check and easy to do. Moreover, it ensures that the solution
+		// is not full of bogus edges that can cause lots of reads to SlashingSpans. Thus, we can
+		// assume that the storage access of this function is always O(|winners|), not
+		// O(|compact.edge_count()|).
+		ensure!(
+			compact_assignments.unique_targets().len() == winners.len(),
+			Error::<T>::OffchainElectionBogusWinnerCount,
+		);
+
+		// Check that the number of presented winners is sane. Most often we have more candidates
+		// than we need. Then it should be `Self::validator_count()`. Else it should be all the
+		// candidates.
+		let snapshot_validators_length = <SnapshotValidators<T>>::decode_len()
+			.map(|l| l as u32)
```

Comment:
map l to l?

Reply:
it's type conversion from usize to u32,  length is an `Option<usize>`

Reply:
I get that. In theory when we can do something as cheap as `unwrap()? as u32`,  using a map and a closure is not a good choice. In practice, possibly the compiler is optimizing both to the same binary. I guess this change also comes from the upstream. So in that case preferring what is there makes the future merge easier.